File: src/main/java/com/yupi/airouter/service/RoutingService.java
package com.yupi.airouter.service;

import com.yupi.airouter.model.entity.Model;

import java.util.List;

/**
 * 路由服务
 *
 * @author <a href="https://codefather.cn">编程导航学习圈</a>
 */
public interface RoutingService {

    /**
     * 根据策略选择模型
     *
     * @param strategyType   策略类型
     * @param modelType      模型类型（可选）
     * @param requestedModel 用户请求的模型标识（固定模型策略使用）
     * @return 选中的模型
     */
    Model selectModel(String strategyType, String modelType, String requestedModel);

    /**
     * 获取 Fallback 模型列表
     *
     * @param strategyType   策略类型
     * @param modelType      模型类型（可选）
     * @param requestedModel 用户请求的模型标识（固定模型策略使用）
     * @return Fallback 模型列表
     */
    List<Model> getFallbackModels(String strategyType, String modelType, String requestedModel);
}

File: go-backend/internal/service/routing_service.go
package service

import (
	"strings"

	"github.com/yupi/airouter/go-backend/internal/constant"
	"github.com/yupi/airouter/go-backend/internal/model/entity"
	"github.com/yupi/airouter/go-backend/internal/repository"
	"github.com/yupi/airouter/go-backend/internal/strategy"
)

const maxRoutingFallbackSize = 3

type RoutingService struct {
	modelRepo   *repository.ModelRepository
	strategyMap map[string]strategy.RoutingStrategy
}

func NewRoutingService(modelRepo *repository.ModelRepository, strategies []strategy.RoutingStrategy) *RoutingService {
	m := make(map[string]strategy.RoutingStrategy, len(strategies))
	for _, item := range strategies {
		m[item.GetStrategyType()] = item
	}
	return &RoutingService{
		modelRepo:   modelRepo,
		strategyMap: m,
	}
}

func (s *RoutingService) SelectModel(strategyType, modelType, requestedModel string) (*entity.Model, []entity.Model, error) {
	finalStrategy := s.getStrategy(strategyType)
	models, err := s.modelRepo.ListRoutable(modelType)
	if err != nil {
		return nil, nil, err
	}
	selected := finalStrategy.SelectModel(models, requestedModel)
	fallbacks := finalStrategy.GetFallbackModels(models, requestedModel)
	if len(fallbacks) > maxRoutingFallbackSize {
		fallbacks = fallbacks[:maxRoutingFallbackSize]
	}
	return selected, fallbacks, nil
}

func (s *RoutingService) DetermineStrategyType(requestedStrategy, requestedModel string) string {
	if strings.TrimSpace(requestedStrategy) != "" {
		return requestedStrategy
	}
	if strings.TrimSpace(requestedModel) != "" {
		return constant.RoutingStrategyFixed
	}
	return constant.RoutingStrategyAuto
}

func (s *RoutingService) getStrategy(strategyType string) strategy.RoutingStrategy {
	if item, ok := s.strategyMap[strategyType]; ok {
		return item
	}
	if autoStrategy, ok := s.strategyMap[constant.RoutingStrategyAuto]; ok {
		return autoStrategy
	}
	return strategy.NewAutoRoutingStrategy()
}
